File: C-Class/__fundamentals/bit_manipulation/number_bit_ones/number_bit_ones.cpp

/*
 *  Programmer:                     Jeff C. Cheng
 *  Last modified:                  11:57AM 02-11-2024
 *  Problem:                        Number of Bits Which Are One (Zero)
 *  @param {Integer} num
 *  @return {Integer}
 */

#include <iostream>

class Solution
{
    public:
        int number_of_bit_ones(int num)
        {
            int sum = 0;

            while (num > 0)
            {

//********************************************************************************************************************
//  Step 1: while number is not 0, count the number of 1's
//********************************************************************************************************************

                num = num & (num-1);
                sum++;
            }

            return sum;
        }
};

int main()
{
    //  8 4 2 0
    //  1 0 0 0
    //  1 0 1 0
    Solution solution;

    int number1 = 8;
    int number2 = 10;

    std::cout << number1 << " has " << solution.number_of_bit_ones(number1) << " bit(s) which are 1." << std::endl;
    std::cout << number2 << " has " << solution.number_of_bit_ones(number2) << " bit(s) which are 1." << std::endl;

    return 0;
}

File: Web/JS/__fundamentals/number-bit-ones/number-bit-ones.js

//
//  *  Programmer:                     Jeff C. Cheng
//  *  Last modified:                  05:36PM 02-11-2024
//  *  Problem:                        Number of Bits Which Are One (Zero)
//  @param {integer} number
//  @return {integer}
//

let numberBitOnes = (number) =>
{

    let sum = 0;

    while (number > 0)
    {

//********************************************************************************************************************
//  Step 1: while number is not 0, count the number of 1's
//********************************************************************************************************************

        number = number & (number-1);
        sum += 1;
    }
    return sum;
}

//  64  32  16  8   4   2   0
//  1   0   1   1   0   0   0
//  0   0   1   0   1   0   0

let number1 = 88;
let number2 = 20;

console.log(number1 + " has " + numberBitOnes(number1) + " bit(s), which are one.");
console.log(number2 + " has " + numberBitOnes(number2) + " bit(s), which are one.");